File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.mini = math.inf
        

    def push(self, val: int) -> None:
        #If stack is empty, then just push and update mini
        if not self.stack:
            self.stack.append(val)
            self.mini = val
        #If stack is not emoty, then check if val < mini?
        #If not, then just push and dont update mini. Else
        #push modified val & update mini
        else:
            if val < self.mini:
                modified_val = (2 * val) - self.mini
                self.stack.append(modified_val)
                self.mini = val
            else:
                self.stack.append(val)

        
    def pop(self) -> None:
        if self.stack:
            #check if stack's top is modified val?
            #If yes then rollback to previous mini and then pop.Otherwose directly pop
            if self.stack[-1] < self.mini:
                self.mini = (2 * self.mini) - self.stack[-1]
    
            self.stack.pop()
        

    def top(self) -> int:
        if self.stack:
             #check if stack's top is modified val?
            #If yes then return mini as top ele. Otherwise return whatever is top ele
            if self.stack[-1] < self.mini:
                return self.mini
            else:
                return self.stack[-1]
        
        return -1

    def getMin(self) -> int:
        if self.stack:
            return self.mini


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Integer> st1;
    Stack<Integer> st2;
    public MinStack() {
        st1 = new Stack<Integer>();
        st2 = new Stack<Integer>();
    }
    
    public void push(int val) {
        st1.push(val);
        if(st2.isEmpty() || st2.peek() >= val){
            st2.push(val);
        }
    }
    
    public void pop() {
        if(st1.peek().equals(st2.peek())){
            st2.pop();
        }
        st1.pop();
    }
    
    public int top() {
        if(!st1.isEmpty()) return st1.peek();

        return -1;
    }
    
    public int getMin() {
        if(!st2.isEmpty()) return st2.peek();

        return -1;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */